// @cerebral/fluent/src/Hoc.ts
/*import React from 'react'
import { View, throwError } from 'cerebral'
import PropTypes from 'prop-types'

class BaseComponent extends React.Component {
  constructor(dependencies, mergeProps, props, controller, name) {
    super(props)
    if (!controller) {
      throwError(
        'Can not find controller, did you remember to use the Container component? Read more at: http://cerebraljs.com/docs/api/components.html#react'
      )
    }

    this.onUpdate = this.onUpdate.bind(this)
    this.view = new View({
      dependencies,
      mergeProps,
      props,
      controller,
      displayName: name,
      onUpdate: this.onUpdate,
    })
  }
 
  componentWillMount() {
    this.view.mount()
  }

  shouldComponentUpdate() {
    return false
  }
 
  componentWillReceiveProps(nextProps) {
    const hasUpdate = this.view.onPropsUpdate(this.props, nextProps)
    if (hasUpdate) {
      this.forceUpdate()
    }
  }

  componentWillUnmount() {
    this.view.unMount()
  }
  onUpdate() {
    this.forceUpdate()
  }
}

export default function HOC(dependencies, mergeProps, Component) {
  class CerebralComponent extends BaseComponent {
    constructor(props, context) {
      super(
        dependencies,
        mergeProps,
        props,
        context.controller,
        Component.displayName || Component.name
      )
    }
    toJSON() {
      return this.view._displayName
    }
    render() {
      return React.createElement(Component, this.view.getProps(this.props))
    }
  }
  CerebralComponent.displayName = `CerebralWrapping_${Component.displayName ||
    Component.name}`

  CerebralComponent.contextTypes = {
    controller: PropTypes.object,
  }

  return CerebralComponent
}
*/
