// cerebral/src/Controller.js
import DependencyStore from './DependencyStore'
import BaseController from './BaseController'
import Model from './Model'
import View from './View'
import { Tag } from 'function-tree'
import { ensureStrictPath } from './utils'

/*
  The controller is where everything is attached. The devtools
  is attached directly. Also a top level module is created.
  The controller creates the function tree that will run all signals,
  based on top level providers and providers defined in modules
*/
class Controller extends BaseController {
  constructor(rootModule, options) {
    super(
      rootModule,
      Object.assign(
        {
          Model,
        },
        options
      )
    )

    this.dependencyStore = new DependencyStore()
    this.flush = this.flush.bind(this)

    this.on('asyncFunction', (execution, funcDetails) => {
      if (!funcDetails.isParallel) {
        this.flush()
      }
    })
    this.on('parallelStart', () => this.flush())
    this.on(
      'parallelProgress',
      (execution, currentPayload, functionsResolving) => {
        if (functionsResolving === 1) {
          this.flush()
        }
      }
    )
    this.on('end', () => this.flush())

    this.getState = this.getState.bind(this)
    this.getSignal = this.getSignal.bind(this)
    this.getSignals = this.getSignals.bind(this)
    this.getComputed = this.getComputed.bind(this)
  }
  /*
    Whenever components needs to be updated, this method
    can be called
  */
  flush(force) {
    const changes = this.model.flush()

    if (!force && !changes.length) {
      return
    }

    this.updateDependencies(changes, force)
    this.emit('flush', changes, Boolean(force))
  }
  updateDependencies(changes, force) {
    let dependenciesToUpdate = []

    if (force) {
      dependenciesToUpdate = this.dependencyStore.getAllUniqueEntities()
    } else {
      dependenciesToUpdate = this.dependencyStore.getUniqueEntities(changes)
    }

    const start = Date.now()
    dependenciesToUpdate.forEach((dependency) => {
      if (this.devtools && dependency instanceof View) {
        this.devtools.updateComponentsMap(dependency)
      }
      dependency.onUpdate(changes, force)
    })
    const end = Date.now()

    if (this.devtools && dependenciesToUpdate.length) {
      const componentsToUpdate = dependenciesToUpdate.filter(
        (dependency) => dependency instanceof View
      )
      if (componentsToUpdate.length) {
        this.devtools.sendComponentsMap(componentsToUpdate, changes, start, end)
      }
    }
  }
  /*
    Create an object used to grab data from tags
  */
  createGetters(props, modulePath = []) {
    const modulePathString = modulePath.length ? modulePath.join('.') + '.' : ''

    return {
      props,
      state: this.getState,
      signal: this.getSignal,
      signals: this.getSignals,
      computed: this.getComputed,
      module: (path) => this.getState(modulePathString + path),
    }
  }
  /*
    Creates a dependency map to be used with the store
  */
  createDependencyMap(dependencies, props, modulePath) {
    const getters = this.createGetters(props, modulePath)

    return dependencies.reduce((currentDepsMap, dependency) => {
      if (
        (dependency instanceof Tag && dependency.type === 'state') ||
        dependency.type === 'module'
      ) {
        return dependency
          .getTags(getters)
          .reduce((updatedCurrentDepsMap, tag) => {
            const path = tag.getPath(getters)
            const strictPath = ensureStrictPath(path, this.getState(path))

            updatedCurrentDepsMap[strictPath] = true

            return updatedCurrentDepsMap
          }, currentDepsMap)
      }
      return currentDepsMap
    }, {})
  }
}

export default Controller
