// cerebral/src/View.js
import { Tag } from 'function-tree'
import {
  getChangedProps,
  throwError,
  createResolver,
  getStateTreeProp,
  noop,
} from './utils'

class View {
  constructor({
    dependencies,
    mergeProps,
    props,
    controller,
    displayName,
    onUpdate,
  }) {
    if (typeof dependencies === 'function') {
      throwError(
        'You can not use a function to define dependencies. Use tags or a function on the specific property you want to dynamically create'
      )
    }

    if (!dependencies) {
      throwError(
        'There is no reason to connect a component that has no dependencies'
      )
    }
    this.mergeProps = mergeProps
    this.controller = controller
    this._displayName = displayName
    this._hasWarnedBigComponent = false
    this.isUnmounted = false
    this.updateComponent = onUpdate || noop
    this.propKeys = Object.keys(props || {})
    this._verifyPropsWarned = false

    if (this.controller.devtools && this.controller.devtools.warnStateProps) {
      this.verifyProps(props)
    }

    this.dependencies = dependencies
    this.dependencyMap = this.controller.createDependencyMap(
      Object.keys(dependencies).map((key) => dependencies[key]),
      props
    )
    this.computedSubscriptions = this.subscribeToComputeds(props)
  }
  /*
    A method to ensure objects and arrays from state tree are not passed as props
  */
  verifyProps(props) {
    const key = getStateTreeProp(props)

    if (key && !this._verifyPropsWarned) {
      console.warn(
        `You are passing an ${
          Array.isArray(props[key]) ? 'array' : 'object'
        } to the component "${
          this._displayName
        }" on prop "${key}" which is from the Cerebral state tree. You should not do this, but rather connect it directly to this component. This will optimize the component and avoid any rerender issues.`
      )
      this._verifyPropsWarned = true
    }
  }
  mount() {
    const depsMap = Object.assign({}, this.dependencyMap)

    this.controller.dependencyStore.addEntity(this, depsMap)

    if (this.controller.devtools) {
      this.controller.devtools.updateComponentsMap(this, depsMap)
    }
  }
  onUpdate(...args) {
    if (this.isUnmounted) {
      return
    }

    this.updateComponent(...args)
  }
  unMount() {
    const depsMap = Object.assign({}, this.dependencyMap)
    this.controller.dependencyStore.removeEntity(this, depsMap)

    if (this.controller.devtools) {
      this.controller.devtools.updateComponentsMap(this, null, depsMap)
    }

    this.isUnmounted = true
  }
  onPropsUpdate(props, nextProps) {
    if (this.controller.devtools) {
      this.verifyProps(nextProps)
    }

    const propsChanges = getChangedProps(props, nextProps)
    if (propsChanges.length) {
      this.updateFromProps(propsChanges, nextProps)

      return true
    }

    return false
  }
  /*
    Called by component when props are passed from parent and they
    have changed. In this situation both tags and depndency trackers might
    be affected. Tags are just updated and dependency trackers are matched
    on props changed
  */
  updateFromProps(propsChanges, props) {
    this.update(props)
  }
  /*
    Called by Container when the components state dependencies
    has changed. In this scenario we need to run any dependencyTrackers
    that matches the state changes. There is no need to update the tags
    as their declared state deps can not change
  */
  updateFromState(stateChanges, props, force) {
    this.update(props)
  }
  /*
    Run update, re-evaluating the tags and computed, if neccessary
  */
  update(props) {
    const previousdependencyMap = this.dependencyMap

    this.dependencyMap = this.controller.createDependencyMap(
      Object.keys(this.dependencies).map((key) => this.dependencies[key]),
      props
    )

    const prevDepsMap = Object.assign({}, previousdependencyMap)
    const nextDepsMap = Object.assign({}, this.dependencyMap)
    this.controller.dependencyStore.updateEntity(this, prevDepsMap, nextDepsMap)

    if (this.controller.devtools) {
      this.controller.devtools.updateComponentsMap(
        this,
        nextDepsMap,
        prevDepsMap
      )
    }
  }
  subscribeToComputeds(props) {
    const getters = this.controller.createGetters(props)
    return Object.keys(this.dependencies).reduce((computeds, key) => {
      if (
        this.dependencies[key] instanceof Tag &&
        this.dependencies[key].type === 'computed'
      ) {
        return computeds.concat(
          this.dependencies[key]
            .getValue(getters)
            .subscribe(this.updateComponent)
        )
      }

      return computeds
    }, [])
  }
  /*
    Runs whenever the component has an update and renders.
    Extracts the actual values from dependency trackers and/or tags
  */
  getProps(props = {}, includeProps = true) {
    const getters = this.controller.createGetters(props)
    const dependenciesProps = Object.keys(this.dependencies).reduce(
      (currentProps, key) => {
        const tag = this.dependencies[key]
        if (tag.type === 'state') {
          const path = tag.getPath(getters)
          const value = getters.state(path)
          if (path.substr(path.length - 2, 2) === '.*') {
            currentProps[key] = value ? Object.keys(value) : []
          } else {
            currentProps[key] = value
          }
        } else if (tag.type === 'computed') {
          currentProps[key] = tag.getValue(getters).getValue(props)
        } else {
          currentProps[key] = tag.getValue(getters)
        }

        return currentProps
      },
      {}
    )

    if (
      this.controller.devtools &&
      this.controller.devtools.bigComponentsWarning &&
      !this._hasWarnedBigComponent &&
      Object.keys(this.dependencies).length >=
        this.controller.devtools.bigComponentsWarning
    ) {
      console.warn(
        `Component named ${
          this._displayName
        } has a lot of dependencies, consider refactoring or adjust this option in devtools`
      )
      this._hasWarnedBigComponent = true
    }

    if (this.mergeProps) {
      return this.mergeProps(
        dependenciesProps,
        props,
        createResolver(this.createTagGetters(props))
      )
    }

    return Object.assign({}, includeProps ? props : {}, dependenciesProps)
  }
}

export default View
