// cerebral/src/Watch.js
class Watch {
  constructor(type) {
    this.id = 'Watch_' + Watch.nextWatchId++
    this.name = null
    this.type = type
    this.executedCount = 0
    this.controller = null
    this.modulePath = ''
    this.subscribers = []
    this.dependencyMap = null
  }
  create(controller, modulePath, name) {
    this.name = name
    this.controller = controller
    this.modulePath = modulePath

    return this
  }
  onUpdate(changes, force) {
    this.executedCount++
    this.subscribers.forEach((subscription) => {
      subscription.cb(changes, force)
    })
  }
  subscribe(cb) {
    if (!this.dependencyMap) {
      this.dependencyMap = this.createDependencyMap()
      this.controller.dependencyStore.addEntity(this, this.dependencyMap)
      if (this.controller.devtools) {
        this.controller.devtools.updateWatchMap(this, this.dependencyMap)
      }
    }

    const index = this.subscribers.length
    const unsubscribe = () => {
      this.subscribers.splice(index, 1)
    }

    this.subscribers.push({
      unsubscribe,
      cb,
    })

    return unsubscribe
  }
  destroy() {
    this.subscribers.forEach((subscription) => subscription.unsubscribe())
    this.subscribers = null
    if (this.dependencyMap) {
      this.controller.dependencyStore.removeEntity(this, this.dependencyMap)

      if (this.controller.devtools) {
        this.controller.devtools.updateWatchMap(this, null, this.dependencyMap)
      }
    }
  }
  toJSON() {
    return {
      id: this.id,
      executedCount: this.executedCount,
      type: this.type,
      name: this.name,
    }
  }
}

Watch.nextWatchId = 0

export default Watch
