// @cerebral/mobx-state-tree/src/index.js
import { BaseControllerClass } from 'cerebral'
import { extractModuleProp } from 'cerebral/internal'
import Model from './Model'
// mobx-state-tree uses extra fields in modules and therefore needs to declare it's own `Module` constructor.
export { Module } from 'cerebral'

class MobxController extends BaseControllerClass {
  constructor(rootModule, options) {
    super(rootModule, options)
    this.provide = function() {
      return {
        store: this.model.state,
        sequences: extractModuleProp(this.module, 'sequences', (sequences) => {
          return Object.keys(sequences).reduce(
            (runableSequences, sequenceKey) => {
              if (sequences[sequenceKey].run) {
                runableSequences[sequenceKey] = sequences[sequenceKey].run
              } else {
                runableSequences[sequenceKey] = sequences[sequenceKey]
              }

              return runableSequences
            },
            {}
          )
        }),
      }
    }
  }
}

export function Controller(rootModule, options = {}) {
  options.Model = Model

  return new MobxController(rootModule, options)
}
