// @cerebral/firebase/src/index.js
import { Provider } from 'cerebral'
import firebase from 'firebase'
import { stopListening } from './helpers'
import signInAnonymouslyService from './signInAnonymously'
import getUserService from './getUser'
import createOnChildAdded from './createOnChildAdded'
import createOnChildRemoved from './createOnChildRemoved'
import createOnChildChanged from './createOnChildChanged'
import createOnValue from './createOnValue'
import createTask from './createTask'
import value from './value'
import push from './push'
import set from './set'
import update from './update'
import put from './put'
import deleteFile from './delete'
import remove from './remove'
import transaction from './transaction'
import createUserWithEmailAndPassword from './createUserWithEmailAndPassword'
import signInWithEmailAndPassword from './signInWithEmailAndPassword'
import signOutService from './signOut'
import signInWithFacebook from './signInWithFacebook'
import signInWithGoogle from './signInWithGoogle'
import signInWithGithub from './signInWithGithub'
import signInWithCustomToken from './signInWithCustomToken'
import deleteUser from './deleteUser'
import sendPasswordResetEmail from './sendPasswordResetEmail'
import linkWithFacebook from './linkWithFacebook'
import linkWithGithub from './linkWithGithub'
import linkWithGoogle from './linkWithGoogle'
import { setOnDisconnect, cancelOnDisconnect } from './onDisconnect'

export { FirebaseProviderError } from './errors'
export { FirebaseProviderAuthenticationError } from './errors'

export default function FirebaseProviderFactory(options = { payload: {} }) {
  firebase.initializeApp(options.config)

  return Provider(
    (context, functionDetails) => ({
      cancelOnDisconnect,
      createUserWithEmailAndPassword,
      delete: deleteFile,
      deleteFile,
      deleteUser,
      getUser: getUserService,
      linkWithFacebook,
      linkWithGithub,
      linkWithGoogle,
      off: stopListening,
      onChildAdded: createOnChildAdded(context),
      onChildChanged: createOnChildChanged(context),
      onChildRemoved: createOnChildRemoved(context),
      onValue: createOnValue(context),
      push,
      put,
      remove,
      sendPasswordResetEmail,
      set,
      setOnDisconnect,
      signInAnonymously: signInAnonymouslyService,
      signInWithCustomToken: signInWithCustomToken,
      signInWithEmailAndPassword,
      signInWithFacebook: signInWithFacebook,
      signInWithGithub: signInWithGithub,
      signInWithGoogle: signInWithGoogle,
      signOut: signOutService,
      task: createTask(options, context, functionDetails.functionIndex),
      transaction,
      update,
      value,
    }),
    {
      cache: !options.sendTaskExecutionDetails,
    }
  )
}
