// @cerebral/forms/src/form.js
import {compute} from 'cerebral'
import runValidation from './utils/runValidation'
import formToJSON from './helpers/formToJSON'
import getInvalidFormFields from './helpers/getInvalidFormFields'
import getFormFields from './helpers/getFormFields'
import {state, props} from 'cerebral/tags'

function createFields (form) {
  return Object.keys(form).reduce((fields, key) => {
    if (form[key] === Object(form[key])) {
      if ('value' in form[key]) {
        fields[key] = new Field(form[key], form)
      } else {
        fields[key] = createFields(form[key], form)
      }
    } else {
      fields[key] = form[key]
    }

    return fields
  }, {})
}

export class Field {
  constructor (field, form) {
    this._form = form
    Object.assign(this, field, {
      isPristine: typeof field.isPristine === 'undefined' ? true : field.isPristine
    })
  }
  _validate () {
    return Object.assign(this, runValidation(this, this._form)).isValid
  }
}

export class Form {
  constructor (form) {
    Object.assign(this, createFields(form))
    this.isValid = this._validate()
  }
  _validate () {
    function validate (obj) {
      return Object.keys(obj).reduce((isValid, field) => {
        if (obj[field] instanceof Field) {
          const isFieldValid = obj[field]._validate()

          return isValid ? isFieldValid : false
        } else if (obj[field] === Object(obj[field])) {
          const areFieldsValid = validate(obj[field])

          return isValid ? areFieldsValid : false
        } else {
          return isValid
        }
      }, true)
    }

    return validate(this)
  }
  toJSON () {
    return formToJSON(this)
  }
  getInvalidFields () {
    return getInvalidFormFields(this)
  }
  getFields () {
    return getFormFields(this)
  }
}

export function computedField (fieldValueTag) {
  return compute(
    fieldValueTag,
    (fieldValue, get) => {
      if (!fieldValue || typeof fieldValue !== 'object') {
        console.warn(`Cerebral Forms - Field value: ${fieldValueTag} did not resolve to an object`)
        return {}
      }
      const formPath = fieldValueTag.getPath({
        state (path) {
          return get(state`${path}`)
        },
        props (path) {
          return get(props`${path}`)
        }
      })
      const formPathArray = formPath.split('.')
      formPathArray.pop()
      const form = get(state`${formPathArray.join('.')}`)
      const field = new Field(fieldValue, form)
      field._validate()
      return field
    }
  )
}

export default function computedForm (formValueTag) {
  return compute(
    formValueTag,
    (formValue) => {
      if (!formValue || typeof formValue !== 'object') {
        console.warn(`Cerebral Forms - Form value: ${formValueTag} did not resolve to an object`)
        return {}
      }
      return new Form(formValue)
    }
  )
}
