// @cerebral/angular/test/TestComponent.ts
import { Component, ChangeDetectionStrategy } from '@angular/core'
import { connect } from '../src/connect'
import { state, signal } from 'cerebral/tags'
import { CerebralComponent } from '../src/CerebralComponent'

@Component({
    selector: 'test-app',
    template: `
        <input id="count" [value]="count">
        <div id="increase" (click)="increase()">Increase</div>
        <div id="decrease" (click)="decrease()">Decrease</div>
        <div id="increaseFreq" (click)="increaseFreq()">Decrease</div>
        <div id="changeSubCompName" (click)="changeSubCompName()">Decrease</div>
        <sub-component id="subComponent" [subCompObj]="subCompObj"></sub-component>
    `,
    changeDetection: ChangeDetectionStrategy.OnPush,
})
@connect({
    count: state`count`,
    subCompObj: state`subCompObj`,
    increase: signal`increase`,
    decrease: signal`decrease`,
    increaseFreq: signal`increaseFreq`,
    changeSubCompName: signal`changeSubCompName`
})
export class TestComponent extends CerebralComponent {
    constructor() {
        super({_view:"view"}, {flush:() => {}}); // Test only code normally see below to setup the component
    }

    // constructor(cdr: ChangeDetectorRef, controller: ControllerService) {
    //     super(cdr, controller)
    // }
}
